### PR TITLE
Extend integration test framework to allow test setup

### PR DESCRIPTION
**Pull request by stack72**
fixes: #4556 

This will allow us to run arbitrary commands to setup our tests
e.g. dotnet publish, or go build etc. The framework will allow steps
to be passed as follows:

```
func TestAccAwsCsLambda(t *testing.T) {
	test := getAWSBase(t).
		With(integration.ProgramTestOptions{
			Dir: path.Join(getCwd(t), "..", "..", "aws-cs-lambda"),
			TestPreSteps: map[string][]string{
				"dotnet": {
					"publish", "DotnetLambda/src/DotnetLambda/",
				},
			},
		})

	integration.ProgramTest(t, &test)
}
```

It will use the name in the map as the binary to use then will have
a list of args that get passed to the binary. We can see it will then
run the steps as follows:

```
[ /pulumi/examples/aws-cs-lambda ] running presteps in /var/folders/pz/y1yk28590l72qy1cmxrfndcw0000gn/T/p-it-demo-mbplo-aws-cs-lam-0993f109-318073945
[ /pulumi/examples/aws-cs-lambda ] **** Invoke '/usr/local/bin/dotnet publish DotnetLambda/src/DotnetLambda/' in '/var/folders/pz/y1yk28590l72qy1cmxrfndcw0000gn/T/p-it-demo-mbplo-aws-cs-lam-0993f109-318073945'
[ /pulumi/examples/aws-cs-lambda ] Wrote output to /var/folders/pz/y1yk28590l72qy1cmxrfndcw0000gn/T/p-it-demo-mbplo-aws-cs-lam-0993f109-318073945/command-output/dotnet-publish.20200504-000103.2b492.log
```

### CODE REVIEW

**Comment by leezen** [comment]
I think this change generally makes sense -- I've at times wanted to do something similar as well. One thing to note though is that there's already a `RunBuild` option, which I think only works in the case of NodeJS tests where it invokes `yarn build`. I think it would make sense to try to make that into a special case of what you're adding here so that it all flows through the same place.

File: pkg/testing/integration/program.go [hunk 1]
```diff
@@ -217,6 +217,10 @@ type ProgramTestOptions struct {
 	// NoParallel will opt the test out of being ran in parallel.
 	NoParallel bool
 
+	// A list of commands to process before the pulumi up. Can be used for things like
+	// preparing a lambda
+	TestPreSteps map[string][]string
```
[the remark this answers is not in the document]
Comment (stack72) [reply pkg/testing/integration/program.go:222]: Sorry for just getting back to this now - you want me just to allow a script to be passed in and execute that?

FWIW, when I initially started this work, there were pain points around the use of PATH based binaries without actually being able to look them up

So passing an entire script would mean we would need to ensure that the user knew that the paths to binaries need to be set

Is this ok with you?

[the remark this answers is not in the document]
Comment (lukehoban) [reply pkg/testing/integration/program.go:222]: I don’t fully understand that specific issue and how impactful it is here. Totally up to you whether you to land as is or expand to allow a callback. Just noting that the later might be a more general solution.

**Comment by pgavlin** [comment]
@stack72 should we keep this open? Is this something we still need?

**Pull request closed without merging**